File: ts/tests/acceptance.test.ts
/**
 * Acceptance tests for any-llm TypeScript implementation.
 * Validates the client against the acceptance test server.
 *
 * Run with: npx tsx tests/acceptance.test.ts
 *
 * Prerequisites:
 * - Acceptance test server running at http://localhost:8080
 * - Start with: uv run acceptance-tests-server/server.py
 */

import { AnyLLM } from "../src/index.js";
import type { ChatCompletion, ChatCompletionChunk } from "../src/index.js";

const BASE_URL = process.env.TEST_SERVER_URL ?? "http://localhost:8080/v1";
const DUMMY_API_KEY = "test-key";

interface ValidationResult {
  passed: boolean;
  errors: Array<{ field: string; message: string }>;
  scenario: string;
}

interface TestRunSummary {
  test_run_id: string;
  total: number;
  passed: number;
  failed: number;
  by_scenario: Record<string, { passed: number; failed: number }>;
}

type CompletionWithValidation = ChatCompletion & {
  _validation?: ValidationResult;
};

const scenarios = {
  basic_completion: {
    model: "test-basic",
    messages: [{ role: "user" as const, content: "Hello" }],
  },
  tool_calls: {
    model: "test-tools",
    messages: [{ role: "user" as const, content: "What's the weather?" }],
    options: {
      tools: [
        {
          type: "function" as const,
          function: {
            name: "get_weather",
            description: "Get weather for a location",
            parameters: {
              type: "object",
              properties: {
                location: { type: "string" },
              },
              required: ["location"],
            },
          },
        },
      ],
    },
  },
  tool_response: {
    model: "test-tool-response",
    messages: [
      { role: "user" as const, content: "What's the weather in Paris?" },
      {
        role: "assistant" as const,
        content: null,
        tool_calls: [
          {
            id: "call_123",
            type: "function" as const,
            function: {
              name: "get_weather",
              arguments: '{"location": "Paris"}',
            },
          },
        ],
      },
      {
        role: "tool" as const,
        content: '{"temperature": "15C"}',
        tool_call_id: "call_123",
      },
    ],
  },
  streaming: {
    model: "test-stream",
    messages: [{ role: "user" as const, content: "Hello" }],
    stream: true,
    options: {
      streamOptions: { include_usage: true },
    },
  },
  structured_output: {
    model: "test-structured",
    messages: [{ role: "user" as const, content: "Give me a person" }],
    options: {
      responseFormat: {
        type: "json_schema" as const,
        json_schema: {
          name: "person",
          schema: {
            type: "object",
            properties: {
              name: { type: "string" },
              age: { type: "integer" },
            },
          },
        },
      },
    },
  },
  multi_turn: {
    model: "test-multi-turn",
    messages: [
      { role: "user" as const, content: "Hello!" },
      { role: "assistant" as const, content: "Hi there!" },
      { role: "user" as const, content: "How are you?" },
    ],
  },
  system_message: {
    model: "test-system",
    messages: [
      { role: "system" as const, content: "You are a helpful assistant." },
      { role: "user" as const, content: "Hello!" },
    ],
  },
  image_content: {
    model: "test-image",
    messages: [
      {
        role: "user" as const,
        content: [
          { type: "text" as const, text: "What's in this image?" },
          {
            type: "image_url" as const,
            image_url: { url: "https://example.com/image.jpg" },
          },
        ],
      },
    ],
  },
  temperature_params: {
    model: "test-params",
    messages: [{ role: "user" as const, content: "Hello" }],
    options: {
      temperature: 0.7,
      topP: 0.9,
      maxTokens: 100,
      presencePenalty: 0.5,
      frequencyPenalty: 0.5,
    },
  },
};

async function createTestRun(testRunId: string): Promise<void> {
  const serverBase = BASE_URL.replace("/v1", "");
  const url = `${serverBase}/v1/test-runs?test_run_id=${encodeURIComponent(testRunId)}&description=TypeScript%20acceptance%20tests`;
  const response = await fetch(url, { method: "POST" });

  if (!response.ok && response.status !== 409) {
    throw new Error(`Failed to create test run: ${response.status}`);
  }
}

async function getTestRunSummary(testRunId: string): Promise<TestRunSummary> {
  const serverBase = BASE_URL.replace("/v1", "");
  const response = await fetch(
    `${serverBase}/v1/test-runs/${encodeURIComponent(testRunId)}/summary`
  );

  if (!response.ok) {
    throw new Error(`Failed to get test run summary: ${response.status}`);
  }

  return response.json();
}

async function runScenario(
  llm: AnyLLM,
  scenarioName: string,
  scenario: (typeof scenarios)[keyof typeof scenarios],
  testRunId: string
): Promise<{ passed: boolean; error?: string }> {
  console.log(`  Running scenario: ${scenarioName}...`);

  try {
    if ("stream" in scenario && scenario.stream) {
      const streamScenario = scenario as typeof scenarios.streaming;
      const chunks: ChatCompletionChunk[] = [];

      for await (const chunk of llm.completionStream(
        streamScenario.model,
        streamScenario.messages,
        streamScenario.options
      )) {
        chunks.push(chunk);
      }

      if (chunks.length === 0) {
        return { passed: false, error: "No chunks received in streaming response" };
      }

      console.log(`    ✓ Received ${chunks.length} chunks`);
      return { passed: true };
    } else {
      const response = (await llm.completion(
        scenario.model,
        scenario.messages,
        "options" in scenario ? scenario.options : undefined
      )) as CompletionWithValidation;

      const validation = response._validation;

      if (validation) {
        if (validation.passed) {
          console.log(`    ✓ Validation passed`);
          return { passed: true };
        } else {
          const errorMsgs = validation.errors
            .map((e) => `${e.field}: ${e.message}`)
            .join("; ");
          console.log(`    ✗ Validation failed: ${errorMsgs}`);
          return { passed: false, error: errorMsgs };
        }
      }

      if (response.choices && response.choices.length > 0) {
        console.log(`    ✓ Response received (no validation in response)`);
        return { passed: true };
      }

      return { passed: false, error: "Invalid response structure" };
    }
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    console.log(`    ✗ Error: ${errorMessage}`);
    return { passed: false, error: errorMessage };
  }
}

async function runProviderTests(
  providerName: string,
  testRunId: string
): Promise<{ passed: number; failed: number; results: Record<string, boolean> }> {
  console.log(`\nTesting provider: ${providerName}`);
  console.log("=".repeat(40));

  let llm: AnyLLM;
  try {
    llm = await AnyLLM.create(providerName, DUMMY_API_KEY, BASE_URL, {
      defaultHeaders: { "X-Test-Run-Id": testRunId },
    });
  } catch (error) {
    console.log(
      `  Failed to initialize provider: ${error instanceof Error ? error.message : error}`
    );
    return {
      passed: 0,
      failed: Object.keys(scenarios).length,
      results: Object.fromEntries(
        Object.keys(scenarios).map((k) => [k, false])
      ),
    };
  }

  let passed = 0;
  let failed = 0;
  const results: Record<string, boolean> = {};

  for (const [scenarioName, scenario] of Object.entries(scenarios)) {
    const result = await runScenario(llm, scenarioName, scenario, testRunId);
    results[scenarioName] = result.passed;
    if (result.passed) {
      passed++;
    } else {
      failed++;
    }
  }

  console.log(`\n  Summary: ${passed} passed, ${failed} failed`);
  return { passed, failed, results };
}

async function main(): Promise<void> {
  const testRunId = `ts-${Date.now()}`;
  console.log("any-llm TypeScript Acceptance Tests");
  console.log("====================================");
  console.log(`Test Run ID: ${testRunId}`);
  console.log(`Server URL: ${BASE_URL}`);

  try {
    await createTestRun(testRunId);
    console.log("Test run created successfully.\n");
  } catch (error) {
    console.error(
      `Warning: Could not create test run: ${error instanceof Error ? error.message : error}`
    );
  }

  const providersToTest = ["openai"];
  const totalResults: Record<
    string,
    { passed: number; failed: number; results: Record<string, boolean> }
  > = {};

  for (const provider of providersToTest) {
    totalResults[provider] = await runProviderTests(provider, testRunId);
  }

  console.log("\n" + "=".repeat(60));
  console.log("FINAL RESULTS");
  console.log("=".repeat(60));

  let totalPassed = 0;
  let totalFailed = 0;

  for (const [provider, result] of Object.entries(totalResults)) {
    console.log(`\n${provider}:`);
    console.log(`  Passed: ${result.passed}`);
    console.log(`  Failed: ${result.failed}`);
    totalPassed += result.passed;
    totalFailed += result.failed;
  }

  console.log(`\nTotal: ${totalPassed} passed, ${totalFailed} failed`);

  try {
    const summary = await getTestRunSummary(testRunId);
    console.log("\nServer-side summary:");
    console.log(`  Total: ${summary.total}`);
    console.log(`  Passed: ${summary.passed}`);
    console.log(`  Failed: ${summary.failed}`);
  } catch {
    console.log("\nNote: Could not fetch server-side summary.");
  }

  if (totalFailed > 0) {
    process.exit(1);
  }
}

main().catch((error) => {
  console.error("Fatal error:", error);
  process.exit(1);
});

File: go/acceptance_test.go
package anyllm_test

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"github.com/stretchr/testify/suite"

	anyllm "github.com/mozilla-ai/any-llm/go"
)

var (
	baseURL     = getEnv("TEST_SERVER_URL", "http://localhost:8080/v1")
	dummyAPIKey = "test-key"
)

func getEnv(key, fallback string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return fallback
}

// TestScenario represents a test scenario from the acceptance test server.
type TestScenario struct {
	Model    string           `json:"model"`
	Messages []anyllm.Message `json:"messages"`
	Stream   bool             `json:"stream,omitempty"`
	Options  map[string]any   `json:"options,omitempty"`
}

// TestRunSummary represents the summary of a test run.
type TestRunSummary struct {
	Total      int            `json:"total"`
	Passed     int            `json:"passed"`
	Failed     int            `json:"failed"`
	ByScenario map[string]any `json:"by_scenario"`
}

// AcceptanceSuite is the test suite for acceptance tests, similar to pytest fixtures.
type AcceptanceSuite struct {
	suite.Suite
	testRunID  string
	serverBase string
	client     anyllm.LLMClient
	scenarios  map[string]TestScenario
}

// SetupSuite runs once before all tests in the suite (like pytest's session-scoped fixtures).
func (s *AcceptanceSuite) SetupSuite() {
	s.testRunID = fmt.Sprintf("go-%d", time.Now().UnixMilli())
	s.serverBase = baseURL[:len(baseURL)-3] // Remove /v1

	// Create test run
	err := s.createTestRun()
	if err != nil {
		s.T().Skipf("Could not create test run (is the acceptance test server running?): %v", err)
	}

	// Load scenarios
	s.scenarios, err = s.loadScenarios()
	require.NoError(s.T(), err, "Failed to load test scenarios")

	// Create client
	s.client, err = anyllm.Create(anyllm.ProviderOpenAI, &anyllm.ClientConfig{
		APIKey:  dummyAPIKey,
		APIBase: baseURL,
		DefaultHeaders: map[string]string{
			"X-Test-Run-Id": s.testRunID,
		},
	})
	require.NoError(s.T(), err, "Failed to create client")
}

// TearDownSuite runs once after all tests complete.
func (s *AcceptanceSuite) TearDownSuite() {
	summary, err := s.getTestRunSummary()
	if err != nil {
		s.T().Logf("Warning: Failed to get test run summary: %v", err)
		return
	}

	assert.Zero(s.T(), summary.Failed, "Test run had %d failures out of %d tests", summary.Failed, summary.Total)
}

// TestScenarios runs all scenario tests dynamically.
func (s *AcceptanceSuite) TestScenarios() {
	for name, scenario := range s.scenarios {
		s.Run(name, func() {
			s.runScenario(scenario)
		})
	}
}

func (s *AcceptanceSuite) runScenario(scenario TestScenario) {
	ctx := context.Background()
	opts := convertOptions(scenario.Options)

	if scenario.Stream {
		s.runStreamingTest(ctx, scenario, opts)
	} else {
		s.runCompletionTest(ctx, scenario, opts)
	}
}

func (s *AcceptanceSuite) runCompletionTest(ctx context.Context, scenario TestScenario, opts *anyllm.CompletionOptions) {
	resp, err := s.client.Completion(ctx, scenario.Model, scenario.Messages, opts)
	require.NoError(s.T(), err, "Completion failed")
	require.NotEmpty(s.T(), resp.Choices, "Response has no choices")

	// Check validation result if present
	if resp.Validation != nil {
		assert.True(s.T(), resp.Validation.Passed,
			"Validation failed for scenario %s: %v", resp.Validation.Scenario, resp.Validation.Errors)
	}
}

func (s *AcceptanceSuite) runStreamingTest(ctx context.Context, scenario TestScenario, opts *anyllm.CompletionOptions) {
	// Ensure stream options are set
	if opts == nil {
		opts = &anyllm.CompletionOptions{}
	}
	includeUsage := true
	opts.StreamOptions = &anyllm.StreamOptions{
		IncludeUsage: &includeUsage,
	}

	chunkChan, errChan := s.client.CompletionStream(ctx, scenario.Model, scenario.Messages, opts)

	var chunks []anyllm.ChatCompletionChunk
	for {
		select {
		case chunk, ok := <-chunkChan:
			if !ok {
				// Channel closed, check for errors
				select {
				case err := <-errChan:
					require.NoError(s.T(), err, "Stream error")
				default:
				}
				goto done
			}
			chunks = append(chunks, chunk)
		case err := <-errChan:
			require.NoError(s.T(), err, "Stream error")
			return
		}
	}

done:
	assert.NotEmpty(s.T(), chunks, "No chunks received in streaming response")
}

func (s *AcceptanceSuite) createTestRun() error {
	url := fmt.Sprintf("%s/v1/test-runs?test_run_id=%s&description=Go%%20acceptance%%20tests", s.serverBase, s.testRunID)
	resp, err := http.Post(url, "application/json", nil)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusConflict {
		return fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}
	return nil
}

func (s *AcceptanceSuite) loadScenarios() (map[string]TestScenario, error) {
	url := fmt.Sprintf("%s/v1/test-data", s.serverBase)
	resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var data struct {
		Scenarios map[string]TestScenario `json:"scenarios"`
	}
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, err
	}

	return data.Scenarios, nil
}

func (s *AcceptanceSuite) getTestRunSummary() (*TestRunSummary, error) {
	url := fmt.Sprintf("%s/v1/test-runs/%s/summary", s.serverBase, s.testRunID)
	resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var summary TestRunSummary
	if err := json.Unmarshal(body, &summary); err != nil {
		return nil, err
	}

	return &summary, nil
}

// TestAcceptance is the entry point for running the acceptance test suite.
func TestAcceptance(t *testing.T) {
	suite.Run(t, new(AcceptanceSuite))
}

func convertOptions(opts map[string]any) *anyllm.CompletionOptions {
	if opts == nil || len(opts) == 0 {
		return nil
	}

	result := &anyllm.CompletionOptions{}

	if v, ok := opts["temperature"].(float64); ok {
		result.Temperature = &v
	}
	if v, ok := opts["top_p"].(float64); ok {
		result.TopP = &v
	}
	if v, ok := opts["max_tokens"].(float64); ok {
		i := int(v)
		result.MaxTokens = &i
	}
	if v, ok := opts["presence_penalty"].(float64); ok {
		result.PresencePenalty = &v
	}
	if v, ok := opts["frequency_penalty"].(float64); ok {
		result.FrequencyPenalty = &v
	}
	if v, ok := opts["tools"].([]any); ok && len(v) > 0 {
		result.Tools = convertTools(v)
	}
	if v, ok := opts["response_format"].(map[string]any); ok {
		result.ResponseFormat = convertResponseFormat(v)
	}
	if v, ok := opts["stream_options"].(map[string]any); ok {
		result.StreamOptions = convertStreamOptions(v)
	}

	return result
}

func convertTools(tools []any) []anyllm.Tool {
	result := make([]anyllm.Tool, 0, len(tools))
	for _, t := range tools {
		if toolMap, ok := t.(map[string]any); ok {
			tool := anyllm.Tool{
				Type: "function",
			}
			if fn, ok := toolMap["function"].(map[string]any); ok {
				tool.Function = anyllm.FunctionDefinition{
					Name:        getStringField(fn, "name"),
					Description: getStringField(fn, "description"),
				}
				if params, ok := fn["parameters"].(map[string]any); ok {
					tool.Function.Parameters = params
				}
			}
			result = append(result, tool)
		}
	}
	return result
}

func convertResponseFormat(rf map[string]any) *anyllm.ResponseFormat {
	result := &anyllm.ResponseFormat{
		Type: getStringField(rf, "type"),
	}
	if js, ok := rf["json_schema"].(map[string]any); ok {
		result.JSONSchema = &anyllm.JSONSchemaFormat{
			Name: getStringField(js, "name"),
		}
		if schema, ok := js["schema"].(map[string]any); ok {
			result.JSONSchema.Schema = schema
		}
	}
	return result
}

func convertStreamOptions(so map[string]any) *anyllm.StreamOptions {
	result := &anyllm.StreamOptions{}
	if v, ok := so["include_usage"].(bool); ok {
		result.IncludeUsage = &v
	}
	return result
}

func getStringField(m map[string]any, key string) string {
	if v, ok := m[key].(string); ok {
		return v
	}
	return ""
}
